qiskit/result/sampled_expval.py

```diff
@@ -0,0 +1,83 @@
+# This code is part of Qiskit.
+#
+# (C) Copyright IBM 2022.
+#
+# This code is licensed under the Apache License, Version 2.0. You may
+# obtain a copy of this license in the LICENSE.txt file in the root directory
+# of this source tree or at http://www.apache.org/licenses/LICENSE-2.0.
+#
+# Any modifications or derivative works of this code must retain this
+# copyright notice, and modified files need to carry a notice indicating
+# that they have been altered from the originals.
+# pylint: disable=cyclic-import
+
+"""Routines for computing expectation values from sampled distributions"""
+import numpy as np
+
+
+# pylint: disable=import-error
+from qiskit._accelerate.sampled_exp_val import sampled_expval_float, sampled_expval_complex
+from qiskit.exceptions import QiskitError
+from .distributions import QuasiDistribution, ProbDistribution
+
+
+# A list of valid diagonal operators
+OPERS = ["Z", "I", "0", "1"]
+
+
+def sampled_expectation_value(dist, oper):
```

Comment:
I'm not sure about this, it could be quite confusing if users search for "expectation value" and will find a function that only works for diagonal operators/distributions -- personally I would not have known that living in the `qiskit.result` namespace means it only supports that 😄 Since this is a more specific function maybe being explicit about the name would be safer?

Reply:
Yeah, that was the idea.

Reply:
Right "result" doesn't give you a clue. I just meant from the PL perspective it is in a name space that only works on  samples. If a user searches for "expectation value" I fear they'll find a hodge podge of methods and functions. They'll have to think about the types of the inputs and select the correct way to do it. This won't be a major problem. But, a unified design, revisiting all the relevant pieces (eg, noted inconsistencies between `Counts` and distributions), would be great. Someone has to find the time and see-it-through energy.